### PR TITLE
Fds/test coverage 3148

### LINKED ISSUE #3148: Add test coverage for Serializable range request conflict detection
See https://github.com/palantir/atlasdb/pull/3147 - we obviously have no test coverage for this and no one has hit this use case, since the code has been broken for ~2 years.

### PR DESCRIPTION
**Pull request by felixdesouza**
**Goals (and why)**:
Fixes #3148. If the fix to #3148 is reverted, the added test will fail. Also cleaned up some things in this class and neighbouring classes.

**Implementation Description (bullets)**:
Opted to use `Multimap` as opposed to a map with a collection value. Cleaner semantics and less null checks, happy to revert if necessary.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a test that reads from a second row. Should we amend all our tests to read from at least 2 rows/cols?

**Concerns (what feedback would you like?)**:
Conversions to `computeIfAbsent`, `Stream`-ifying, etc.

**Where should we start reviewing?**:
`AbstractSerializationTransactionTest`

**Priority (whenever / two weeks / yesterday)**:
Whenever.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3535)

### CODE REVIEW

**Comment by felixdesouza** [comment]
[no release notes]

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransaction.java [hunk 1]
```diff
@@ -253,23 +256,13 @@ public SerializableTransaction(MetricsManager metricsManager,
...
 
     private ConcurrentNavigableMap<Cell, byte[]> getReadsForTable(TableReference table) {
-        ConcurrentNavigableMap<Cell, byte[]> reads = readsByTable.get(table);
-        if (reads == null) {
-            ConcurrentNavigableMap<Cell, byte[]> newMap = new ConcurrentSkipListMap<>();
-            readsByTable.putIfAbsent(table, newMap);
-            reads = readsByTable.get(table);
-        }
-        return reads;
+        return readsByTable.computeIfAbsent(table, $ -> new ConcurrentSkipListMap<>());
```
Comment (j-baker) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransaction.java]: probably best to not use '$' as a variable name; it's disallowed by our checkstyle but unenforced due to a checkstyle bug. Rather, prefer something like 'unused'.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 2]
```diff
@@ -1460,14 +1446,7 @@ protected void throwIfReadWriteConflictForSerializable(long commitTimestamp) {
...
     private boolean hasWrites() {
-        boolean hasWrites = false;
-        for (SortedMap<?, ?> map : writesByTable.values()) {
-            if (!map.isEmpty()) {
-                hasWrites = true;
-                break;
-            }
-        }
-        return hasWrites;
+        return writesByTable.values().stream().anyMatch(writesForTable -> !writesForTable.isEmpty());
```
Comment (felixdesouza) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: are we opposed to doing `hasWrites` as a `volatile boolean` similar to `hasReads`?

Comment (sandorw) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: Not opposed if you think that's clearer, although this is also sufficient. There's already a lot of state being tracked in this class.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java [hunk 3]
```diff
@@ -702,6 +702,25 @@ public void testMultipleReadsToSameColumnRange() {
         t1.commit();
     }
 
+    @Test
+    public void testMultipleReadsToSameColumnRangeAcrossRows() {
+        byte[] row1 = PtBytes.toBytes("row1");
+        byte[] row2 = PtBytes.toBytes("row2");
+
+        Transaction transaction = startTransaction();
+        Map<byte[], BatchingVisitable<Map.Entry<Cell, byte[]>>> columnRange =
```
Comment (sandorw) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java]: Share the `BatchColumnRangeSelection` to make it clearer it's the same and prevent someone from breaking the test by updating one and not the other.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 2 above)
Comment (felixdesouza) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: I shall leave it, agree with the state tracking bit.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java (hunk 3 above)
Comment (felixdesouza) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java]: good catch

**Pull request merged**